rust/src/vacuum.rs

```diff
@@ -0,0 +1,217 @@
+                source: err,
+            })),
+        }?;
+
+        Ok(VacuumMetrics {
+            files_deleted,
+            dry_run: false,
+        })
+    }
+}
+
+/// Determine which files can be deleted. Does not actually peform the deletion
+pub async fn create_vacuum_plan(
+    table: &DeltaTable,
+    params: Vacuum,
+) -> Result<VacuumPlan, VacuumError> {
+    let min_retention = Duration::milliseconds(table.state.tombstone_retention_millis());
+    let retention_period = params.retention_period.unwrap_or(min_retention);
+    let enforce_retention_duration = params.enforce_retention_duration;
+
+    if enforce_retention_duration && retention_period < min_retention {
+        return Err(VacuumError::InvalidVacuumRetentionPeriod {
+            provided: retention_period.num_hours(),
+            min: min_retention.num_hours(),
+        });
+    }
+
+    let now_millis = Utc::now().timestamp_millis();
+    let expired_tombstones = get_stale_files(table, retention_period, now_millis)?;
+    let valid_files = table.get_file_set();
+
+    let mut files_to_delete = vec![];
+    let mut all_files = table
+        .storage
+        .list_objs(&table.table_uri)
+        .await
+        .map_err(DeltaTableError::from)?;
+
+    while let Some(obj_meta) = all_files.next().await {
+        let obj_meta = obj_meta.map_err(DeltaTableError::from)?;
+        let rel_path = extract_rel_path(&table.table_uri, &obj_meta.path)?;
+
+        if valid_files.contains(rel_path) // file is still being tracked in table
+            || !expired_tombstones.contains(rel_path) // file is not an expired tombstone
+            || is_hidden_directory(table, rel_path)?
+        {
+            continue;
+        }
+
+        files_to_delete.push(obj_meta.path);
+    }
+
+    Ok(VacuumPlan { files_to_delete })
+}
+
+/// Methods to specify various vacuum options and to execute the operation
+impl Vacuum {
+    /// Override the default rention period for which files are deleted.
+    pub fn with_retention_period(mut self, retention_period: Duration) -> Self {
+        self.retention_period = Some(retention_period);
+        self
+    }
+
+    /// Only determine which files should be deleted
+    pub fn dry_run(mut self, dry_run: bool) -> Self {
+        self.dry_run = dry_run;
+        self
+    }
+
+    /// Check if the specified retention period is less than the table's minimum
+    pub fn enforce_retention_duration(mut self, enforce: bool) -> Self {
+        self.enforce_retention_duration = enforce;
+        self
+    }
+
+    /// Perform the vacuum. Returns metrics on which files were deleted
+    pub async fn execute(self, table: &mut DeltaTable) -> Result<VacuumMetrics, VacuumError> {
+        let dry_run = self.dry_run;
+        let plan = create_vacuum_plan(table, self).await?;
+        if dry_run {
+            return Ok(VacuumMetrics {
+                files_deleted: plan.files_to_delete,
+                dry_run: true,
+            });
```

Comment:
excellent - returning metrics like this is exactly the right thing to do. In  #673 this is what's needed to do proper conflict resolution. There we need to know all files read by a transaction. There is a `CurrentTransactionInfo` struct in that PR that holds all we want to know about the current transaction. I see now that this does not care about files removed from the table. but while the complete transaction info object is maybe not suitable here, maybe we want a more generic Metrics struct, that can also report on other metrics?

Reply:
I'm not sure if a generic Metrics struct would really work in Rust but I think it would be a good place for traits. For example, tracking the time an operation took is common so a `TimeCostMetric` trait could be used to obtain that. Ditto for conflict resolution. A trait that specifies all the information required can be defined and then each operation is required to implement it.